### DISCUSSION TITLE
Index type signature

### DISCUSSION THREAD

**Discussion by lucasyarid** [prisma/prisma#4397] [Q&A]
I'm currently having a hard time figuring out how to pass an index to access a prisma value in typescript:

```js
import { PrismaClient } from '@prisma/client'

const findPrismaByIndex = async ({ database, value }: { database: PrismaClient, value }): Promise<boolean> => {
  const found = await prisma[database].findOne({ where: { value } })
  ...
}
```

ts compiler gives me this error:
```
Type 'PrismaClient<PrismaClientOptions, never>' cannot be used as an index type.ts(2538)
```

Reading ts docs seems like one needs to set it up at type/interface declaration like so:
```js
type Users = {
   [email: string]: string,
   ...
}; 
```

But looking at the prisma generated model (`node_modules/.prisma/client/index.d.ts`) it seems like it is not implemented:
```js
export type Users = {
  email: string
  ...
}
```

Is that even possible or remotely non-problematic?
I'm trying to set a helper function to find database uniques so I don't need to create logic for each case.

BTW prisma going great so far!
Thanks

**Comment by lucasyarid** [comment]
Well for now I'm just doing it with `any`, which is not ideal:

```js
const findPrismaByIndex = async ({ database, value }: { database: PrismaClient, value }): Promise<boolean> => {
  const found = !await (prisma as any)[database].findOne(...)
  ...
}
```

**Reply by m10rten** [reply]
So this seems the way to go, but to make TS and your development experience beter, I would make it something like: `Record<string, any>`.

**Reply by m10rten** [reply]
To have something like this: 
```js
export const handlePrismaClient = async (client: PrismaClient, dbUrl: string | undefined) => {
  const conn = await client.$connect();
  console.log(conn);
  console.log(dbUrl);
  const schemas = Object.keys(client).filter((key) => false === key.startsWith("_") && false === key.startsWith("$"));
  console.log(schemas);
  const results = await Promise.all(
    schemas.map(async (schema: string) => {
      const result: Array<any> = await (client as Record<string, any>)[schema].findMany();
      return result;
    }),
  );
  console.log(results);
  return client;
};
```

All it does for you as a developer is skip the typechecking more or less for this line.
But you still have the `:PrismaClient` type for the client.

**Comment by hdngr** [comment]
What is the best way to do this?  I have custom resolver methods in my GQL layer that use the same prisma methods across different methods...  Same as above: 
`prisma[someResource].findUnique()`
```
 No index signature with a parameter of type 'string' was found on type 'PrismaClient<PrismaClientOptions, never>'.
```
Is there a better way to do this with typescript?

Thanks!

**Comment by d2vin** [comment]
I'm also having a similar issue, but my code still runs nonetheless. No disable line error quick fixes popping up on vs code though.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
